File: mobile/src/models/greenlightStatus.ts
import { Model, attribute as attr, Base } from './base'
import { User } from './user'
import moment, { Moment } from 'moment'

export enum GREENLIGHT_STATUSES {
  RED,
  YELLOW,
  GREEN,
  ABSENT,
  UNKNOWN
}

@Model({ singular: 'location', plural: 'locations' })
export class GreenlightStatus extends Base {
  @attr() id: number = -2
  @attr({ type: 'User' }) user: User | null = null
  @attr() status: GREENLIGHT_STATUSES = GREENLIGHT_STATUSES.UNKNOWN

  @attr({ transform: moment })
  statusSetAt: Moment | null = null
  @attr({ transform: moment })
  statusExpiresAt: Moment | null = null
}

File: api/src/main/kotlin/greenlight/entities/GreenlightStatus.kt
package greenlight.entities

import io.micronaut.core.annotation.Introspected
import java.time.Instant
import java.util.*
import javax.persistence.Entity
import javax.persistence.PrePersist
import javax.persistence.Table
import javax.validation.constraints.NotBlank






@Introspected
@Entity
@Table(name = "greenlight_statuses")
class GreenlightStatus : AbstractBase() {
    enum class Statuses(val status: String) {
        RED("red"),
        YELLOW("yellow"),
        GREEN("green"),
        ABSENT("green"),
        UNKNOWN("unknown"),
    }

    var userId : UUID? = null
    var locationId : UUID? = null
    @NotBlank
    var status : String? = null
    @NotBlank
    var statusSetAt : Instant? = null
    @NotBlank
    var statusExpiresAt : Instant? = null
    @NotBlank
    var isOverride : Boolean? = null
    var createdByUserId : UUID? = null
    @NotBlank
    var createdAt : Instant? = null

    @PrePersist
    fun onPersist() {
        createdAt = Instant.now()
    }

}
